src/lib/Output/Generator/AbstractFieldTypeHashGenerator.php

```diff
@@ -0,0 +1,141 @@
+<?php
+
+/**
+ * @copyright Copyright (C) Ibexa AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+
+namespace Ibexa\Rest\Output\Generator;
+
+use Ibexa\Rest\Output\Generator\Data\DataObjectInterface;
+use Ibexa\Rest\Output\Generator\Json\JsonObject;
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\Serializer\Exception\ExceptionInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerInterface;
+
+abstract class AbstractFieldTypeHashGenerator implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    private NormalizerInterface $normalizer;
+
+    private bool $strictMode;
+
+    public function __construct(
+        NormalizerInterface $normalizer,
+        ?LoggerInterface $logger = null,
+        bool $strictMode = false
+    ) {
+        $this->normalizer = $normalizer;
+        $this->logger = $logger ?? new NullLogger();
+        $this->strictMode = $strictMode;
+    }
+
+    /**
+     * Generates the field type value $hashValue as a child of the given Object
+     * using $hashElementName as the property name.
+     */
+    public function generateHashValue(
+        DataObjectInterface $parent,
+        string $hashElementName,
+        mixed $hashValue
+    ): void {
+        $parent->$hashElementName = $this->generateValue($parent, $hashValue);
+    }
+
+    /**
+     * Generates and returns a JSON structure (array or object) depending on $value type
+     * with $parent.
+     *
+     * If $type only contains numeric keys, the resulting structure will be an
+     * JSON array, otherwise a JSON object
+     *
+     * @param array<mixed> $value
+     */
+    protected function generateArrayValue(
+        DataObjectInterface $parent,
+        array $value,
+    ): DataObjectInterface {
+        if ($this->isNumericArray($value)) {
+            return $this->generateListArray($parent, $value);
+        } else {
+            return $this->generateHashArray($parent, $value);
+        }
+    }
+
+    /**
+     * Generates and returns a value based on $hashValue type, with $parent (
+     * if the type of $hashValue supports it).
+     */
+    abstract protected function generateValue(DataObjectInterface $parent, mixed $value): mixed;
+
+    /**
+     * Checks if the given $value is a purely numeric array.
+     *
+     * @param array<mixed> $value
+     */
+    protected function isNumericArray(array $value): bool
+    {
+        foreach (array_keys($value) as $key) {
+            if (is_string($key)) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    protected function generateObjectValue(DataObjectInterface $parent, object $value): mixed
```

Comment:
```suggestion
    /**
     * @throws \Symfony\Component\Serializer\Exception\ExceptionInterface
     */
    protected function generateObjectValue(DataObjectInterface $parent, object $value): mixed
```

Reply:
I'm not sure there is a point in referencing these exceptions, `NormalizerInterface` is annotated with these:
```php
     * @throws InvalidArgumentException   Occurs when the object given is not a supported type for the normalizer
     * @throws CircularReferenceException Occurs when the normalizer detects a circular reference when no circular
     *                                    reference handler can fix it
     * @throws LogicException             Occurs when the normalizer is not called in an expected context
     * @throws ExceptionInterface         Occurs for all the other cases of errors
```

so technically speaking we would have to annotate them pretty much everywhere.

Reply:
> Annotated? How?

Simply put, they are declared on the interface that this class implements. You're asking to repeat it.

Maybe this is a valid location for an `@inheritdoc`.

Reply:
I must admit I'm a bit confused. We usually don't use `@inheritdoc` when implementing an interface as it's self-explanatory for IDE or else, that this direct implementation throws an exception from the implemented interface. Why would we do it now?

Reply:
We don't do it when it is the only thing that PHPDoc says. Because then it's just adding whitespace/noise without any value (which was the case majority of the time).

If, however, this makes PHPStorm or any other tool recognize that we're not handling exception properly, then there is value in it. While PHPStorm is by no means a definite indicator, I know for a fact that PHPStan has the capacity to track exception throwing - we're just not using it yet.

I'd leave this for you and @alongosz to explore, and decide how we should proceed with documenting in this case. If it changes how tools behave, then maybe it becomes worthwile.